Refresh.GameServer/Importing/Mip/MipEncoder.cs

```diff
@@ -0,0 +1,121 @@
+using SixLabors.ImageSharp;
+using SixLabors.ImageSharp.Formats;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing;
+using SixLabors.ImageSharp.Processing.Processors.Dithering;
+using SixLabors.ImageSharp.Processing.Processors.Quantization;
+
+namespace Refresh.GameServer.Importing.Mip;
+
+public class MipEncoder : ImageEncoder
+{
+    protected override void Encode<TPixel>(Image<TPixel> origImage, Stream stream, CancellationToken cancellationToken)
+    {
+        Image<Rgba32> image = origImage.CloneAs<Rgba32>();
+        //Quantize the image into 256 colors
+        image.Mutate(ctx => ctx.Quantize(new WuQuantizer(new QuantizerOptions
+        {
+            Dither = ErrorDither.Atkinson,
+            MaxColors = 256,
+        })));
+        image.SaveAsPng("rawquant.png");
+        Dictionary<Rgba32, byte> outputPalette = new(256);
+        {
+            for (int x = 0; x < image.Width; x++)
+            {
+                for (int y = 0; y < image.Height; y++)
+                {
+                    Rgba32 col = image[x, y];
+                    //If the color is not in the palette
+                    if (!outputPalette.TryGetValue(col, out _))
+                    {
+                        //Add the color to the palette
+                        outputPalette[col] = (byte)outputPalette.Count;
+                    }
+                }
+            }
+        }
+
+        if (outputPalette.Count > 256) throw new Exception("Too many colours in image!");
+
+        bool alpha = outputPalette.Any(col => col.Key.A != 255);
+        
+        MipHeader header = new()
+        {
+            Width = (uint)image.Width,
+            Height = (uint)image.Height,
+            Bpp = 8,
+            NumBlocks = 2,
+            TexMode = 1,
+            Alpha = alpha,
+            Clut = new Rgba32[256],
+        };
+        
+        //Fill in the header's CLUT
+        foreach ((Rgba32 key, byte value) in outputPalette) header.Clut[value] = key;
+
+        //Write the header
+        header.Write(stream);
+        
+        BinaryWriter writer = new(stream);
+
+        {
```

Comment:
Scope is there because C# is silly and says that the `x` and `y` shadow the `x` and `y` used above while getting the palette (even though its in its own scope), i'll move the scope to a function